syntax = "proto3";

package grapevine;

message Message {
    string raw = 1;
    string text = 2;
    string language = 3;
    int64 created_at = 4;
    repeated string urls = 5;
}

service Classifier {
    rpc Classify(Message) returns (Classification) {}
}

message Classification {
    string domain = 1;
    string prediction = 2;
    double confidence = 3;
    string model = 4;
    string version = 5;
    Meta meta = 6;
}

message Meta {
    repeated Sentence sentences = 1;
    SpamAssassinReport spamassassin_report = 2;
    repeated Ask asks = 3;
}

message Sentence {
    double sentence_score = 1;
    repeated double word_scores = 2;
    repeated string words = 3;
}

message SpamAssassinReport {
    repeated SpamAssassinHit quantitative_hits = 1;
    repeated SpamAssassinHit qualitative_hits = 2;
}

message SpamAssassinHit {
    double score = 1;
    string symbol = 2;
    string description = 3;
}

message Ask {
    string sentence = 1;
    string raw_sentence = 2;
    int32 start_idx = 3;
    int32 end_idx = 4;
    string prediction = 5;
    double confidence = 6;
}

service Extractor {
    rpc Extract(Message) returns (Extraction) {}
}

message Extraction {
    string key = 1;
    repeated string values = 2;
    double confidence = 3;
    string model = 4;
    string version = 5;
}